src/transactions/transaction_db.rs

```diff
@@ -0,0 +1,778 @@
+    ) -> Result<*mut ffi::rocksdb_column_family_handle_t, Error> {
+        let cf_name = if let Ok(c) = CString::new(name.as_bytes()) {
+            c
+        } else {
+            return Err(Error::new(
+                "Failed to convert path to CString when creating cf".to_owned(),
+            ));
+        };
+        Ok(unsafe {
+            ffi_try!(ffi::rocksdb_transactiondb_create_column_family(
+                self.inner,
+                opts.inner,
+                cf_name.as_ptr(),
+            ))
+        })
+    }
+
+    pub fn list_cf<P: AsRef<Path>>(opts: &Options, path: P) -> Result<Vec<String>, Error> {
+        DB::list_cf(opts, path)
+    }
+
+    pub fn destroy<P: AsRef<Path>>(opts: &Options, path: P) -> Result<(), Error> {
+        DB::destroy(opts, path)
+    }
+
+    pub fn repair<P: AsRef<Path>>(opts: &Options, path: P) -> Result<(), Error> {
+        DB::repair(opts, path)
+    }
+
+    pub fn path(&self) -> &Path {
+        self.path.as_path()
+    }
+
+    /// Creates a transaction with default options.
+    pub fn transaction(&self) -> Transaction<Self> {
+        self.transaction_opt(&WriteOptions::default(), &TransactionOptions::default())
+    }
+
+    /// Creates a transaction with options.
+    pub fn transaction_opt<'a>(
+        &'a self,
+        write_opts: &WriteOptions,
+        txn_opts: &TransactionOptions,
+    ) -> Transaction<'a, Self> {
+        Transaction {
+            inner: unsafe {
+                ffi::rocksdb_transaction_begin(
+                    self.inner,
+                    write_opts.inner,
+                    txn_opts.inner,
+                    std::ptr::null_mut(),
+                )
+            },
+            _marker: PhantomData::default(),
+        }
+    }
+
+    /// Returns the bytes associated with a key value.
+    pub fn get<K: AsRef<[u8]>>(&self, key: K) -> Result<Option<Vec<u8>>, Error> {
+        self.get_opt(key, &ReadOptions::default())
+    }
+
+    /// Returns the bytes associated with a key value and the given column family.
+    pub fn get_cf<K: AsRef<[u8]>>(
+        &self,
+        cf: &impl AsColumnFamilyRef,
+        key: K,
+    ) -> Result<Option<Vec<u8>>, Error> {
+        self.get_cf_opt(cf, key, &ReadOptions::default())
+    }
+
+    /// Returns the bytes associated with a key value with read options.
+    pub fn get_opt<K: AsRef<[u8]>>(
+        &self,
+        key: K,
+        readopts: &ReadOptions,
+    ) -> Result<Option<Vec<u8>>, Error> {
+        unsafe {
+            let mut val_len: usize = 0;
+            let val_ptr = ffi_try!(ffi::rocksdb_transactiondb_get(
+                self.inner,
+                readopts.inner,
+                key.as_ref().as_ptr() as *const c_char,
+                key.as_ref().len() as size_t,
+                &mut val_len,
+            ));
+            if val_ptr.is_null() {
+                Ok(None)
+            } else {
+                let val = Vec::from_raw_parts(val_ptr as *mut u8, val_len, val_len);
```

Comment:
I realized that I misunderstood the use of `Vec::from_raw_parts`. It's not safe to build Vec from memory allocated in C.

We may need to copy data or build a new type such as `CVec` to wrap C memory.

Reply:
Previously there was `DBVector` implementation, which was removed in:
https://github.com/rust-rocksdb/rust-rocksdb/pull/345/files#diff-7d7d3a6ce5c2a3e8b39ce4f49b78a71365c8e06c0d9ddb78bc2dca68bfeb4a5bL2028

I think it should be reverted, because not all of rocksdb functions use pinnable slice. (especially transaction related)

But if this type would be reverted, make sure that in drop it would use
`rocksdb_free` instead of `libc::free(self.base as *mut c_void);`

Reply:
It's better to support `get_pinned` in rocksdb C API, see facebook/rocksdb#9252.